File: frontend/src/services/answer.ts
import api from './api';

// 答案状态类型
export type AnswerStatus = 'draft' | 'submitted' | 'completed' | 'failed';

// 类型定义
export interface Answer {
  id: number;
  question_set_id: number;
  student_id: number;
  course_id: number;
  student_answers: Record<string, string | string[]> | null;
  grading_results: GradingResults | null;
  total_score: number | null;
  status: AnswerStatus;
  error_message: string | null;
  saved_at: string;
  submitted_at: string | null;
}

export interface GradingResults {
  [key: string]: QuestionGradingResult;
}

export interface QuestionGradingResult {
  question_id: string;
  question_type: string;
  score: number;
  max_score: number;
  feedback: string;
  correct_answer?: string | null;
  analysis?: string;
}

export interface SaveDraftRequest {
  question_set_id: number;
  student_answers: Record<string, string | string[]>;
}

export interface SubmitAnswerRequest {
  question_set_id: number;
  student_answers: Record<string, string | string[]>;
}

// 教师端类型
export interface AnswerStudentInfo {
  id: number;
  username: string;
  full_name: string;
  email: string;
}

export interface TeacherAnswer extends Answer {
  student: AnswerStudentInfo;
}

export interface TeacherScoreUpdate {
  total_score: number;
  question_scores?: Record<string, number>;
}

// 学生端答案 API
export const answerApi = {
  // 保存草稿
  saveDraft: (data: SaveDraftRequest) =>
    api.post<Answer>('/answer/save-draft', data),

  // 提交答案
  submit: (data: SubmitAnswerRequest) =>
    api.post<Answer>('/answer/submit', data),

  // 获取答案详情
  getById: (answerId: number) =>
    api.get<Answer>(`/answer/${answerId}`),

  // 获取学生对某试题集的答案
  getByQuestionSet: (questionSetId: number) =>
    api.get<Answer | null>(`/answer/question-set/${questionSetId}`),
};

// 教师端答案 API
export const teacherAnswerApi = {
  // 获取试题集的所有学生答案
  getByQuestionSet: (questionSetId: number) =>
    api.get<TeacherAnswer[]>(`/answer/teacher/question-set/${questionSetId}`),

  // 获取单份答案详情
  getById: (answerId: number) =>
    api.get<TeacherAnswer>(`/answer/teacher/${answerId}`),

  // 修改分数
  updateScore: (answerId: number, data: TeacherScoreUpdate) =>
    api.patch(`/answer/teacher/${answerId}/score`, data),
};

File: backend/app/schemas/answer.py
"""答案相关的请求/响应模型"""

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field

from backend.app.models.answer import AnswerStatus


class AnswerSaveDraft(BaseModel):
    """保存草稿请求"""

    question_set_id: int
    student_answers: dict[str, Any] = Field(default_factory=dict)


class AnswerCreate(BaseModel):
    """提交答案请求"""

    question_set_id: int
    student_answers: dict[str, Any] = Field(default_factory=dict)


class GradingResult(BaseModel):
    """单题批改结果"""

    question_id: str
    question_type: str
    score: float
    max_score: float
    feedback: str
    correct_answer: str | None = None


class AnswerResponse(BaseModel):
    """答案响应"""

    id: int
    question_set_id: int
    student_id: int
    course_id: int
    status: AnswerStatus
    total_score: float | None
    saved_at: datetime
    submitted_at: datetime | None

    model_config = {"from_attributes": True}


class AnswerDetail(AnswerResponse):
    """答案详情响应"""

    student_answers: dict[str, Any] | None
    grading_results: dict[str, Any] | None
    error_message: str | None


class StudentInfo(BaseModel):
    """学生基本信息"""

    id: int
    username: str
    full_name: str
    email: str


class TeacherAnswerDetail(BaseModel):
    """教师查看的答案详情"""

    id: int
    question_set_id: int
    student_id: int | None = None
    course_id: int
    status: str
    total_score: float | None
    saved_at: datetime
    submitted_at: datetime | None
    student_answers: dict[str, Any] | None
    grading_results: dict[str, Any] | None
    error_message: str | None
    student: StudentInfo

    model_config = {"from_attributes": True}


class TeacherScoreUpdate(BaseModel):
    """教师修改分数请求"""

    total_score: float = Field(..., ge=0, le=100)
    question_scores: dict[str, float] | None = None  # 可选，单题分数
